### PR TITLE
SSL4EO-L: add new dataset

### PR DESCRIPTION
**Pull request by adamjstewart**
Remaining TODOs:

* Add data module
* Add download/extraction support

etm_toa
tm_sr
oli_tirs_toa
oli_sr

### CODE REVIEW

File: tests/datasets/test_ssl4eo.py [hunk 1]
```diff
@@ -14,19 +13,56 @@
...
+        dataset.plot(sample, show_titles=False)
+        plt.close()
 
 
 class TestSSL4EOS12:
-    @pytest.fixture(params=zip(SSL4EOS12.metadata.keys(), [1, 1, 2]))
+    @pytest.fixture(params=zip(SSL4EOS12.metadata.keys(), [1, 2, 4]))
     def dataset(self, monkeypatch: MonkeyPatch, request: SubRequest) -> SSL4EOS12:
-        monkeypatch.setitem(
+        monkeypatch.setitem(  # type: ignore[arg-type]
```
Comment (adamjstewart) [inline tests/datasets/test_ssl4eo.py:59]: * https://github.com/pytest-dev/pytest/issues/10999
* https://github.com/pytest-dev/pytest/pull/11000

File: torchgeo/datasets/ssl4eo.py [hunk 2]
```diff
@@ -17,13 +17,187 @@
...
+        """
+        assert split in self.metadata
+        assert seasons in range(1, 5)
+
+        self.root = root
+        self.split = split
+        self.seasons = seasons
+        self.transforms = transforms
+
+        self.scenes = sorted(os.listdir(root))
```
Comment (adamjstewart) [inline torchgeo/datasets/ssl4eo.py:121]: Will replace this with integer indexing after we remap the folder names.

File: torchgeo/datasets/ssl4eo.py [hunk 3]
```diff
@@ -51,7 +225,10 @@ class SSL4EOS12(NonGeoDataset):
 
     size = 264
 
-    metadata = {
+    _Metadata = TypedDict(
+        "_Metadata", {"filename": str, "md5": str, "bands": list[str]}
+    )
```
Comment (adamjstewart) [inline torchgeo/datasets/ssl4eo.py:230]: TypedDict allows us to remove all uses of Any and cast! We should try to use this in more places.

File: torchgeo/datasets/ssl4eo.py [hunk 4]
```diff
@@ -121,7 +298,7 @@ def __init__(
             RuntimeError: if dataset is missing or checksum fails
         """
         assert split in self.metadata
-        assert seasons in range(4)
+        assert seasons in range(1, 5)
```
Comment (adamjstewart) [inline torchgeo/datasets/ssl4eo.py:301]: This was a bug, oops

**Review by calebrob6 — APPROVED** [review]

**Pull request merged**